calico/felix/futils.py

```diff
@@ -95,6 +99,122 @@ def call_silent(args):
+        assert startupinfo is None and creationflags == 0
+
+        log.debug("Pipes: p2c %s, %s; c2p %s, %s; err %s, %s",
+                  p2cread, p2cwrite,
+                  c2pread, c2pwrite,
+                  errread, errwrite)
+
+        if isinstance(args, types.StringTypes):
+            args = [args]
+        else:
+            args = [a.encode("ascii") for a in args]
+
+        if shell:
+            args = ["/bin/sh", "-c"] + args
+            if executable:
+                args[0] = executable
+
+        if executable is None:
+            executable = args[0]
+
+        self._loop.install_sigchld()
+
+        file_actions = FileActions()
+        # Close parent's pipe ends
+        if p2cwrite is not None:
+            file_actions.add_close(p2cwrite)
+        if c2pread is not None:
+            file_actions.add_close(c2pread)
+        if errread is not None:
+            file_actions.add_close(errread)
+
+        # When duping fds, if there arises a situation
+        # where one of the fds is either 0, 1 or 2, it
+        # is possible that it is overwritten (#12607).
+        fds_to_close_in_parent = []
+        if c2pwrite == 0:
+            c2pwrite = os.dup(c2pwrite)
+            fds_to_close_in_parent.append(c2pwrite)
+        if errwrite == 0 or errwrite == 1:
+            errwrite = os.dup(errwrite)
+            fds_to_close_in_parent.append(errwrite)
+
+        # Dup fds for child
+        def _dup2(a, b):
+            if a is None:
+                a = b
+            # Need to take a dup so we can remove the non-blocking flag
+            a_dup = os.dup(a)
+            log.debug("Duped %s as %s", a, a_dup)
+            fds_to_close_in_parent.append(a_dup)
+            self._remove_nonblock_flag(a_dup)
+            file_actions.add_dup2(a_dup, b)
+        _dup2(p2cread, 0)
+        _dup2(c2pwrite, 1)
+        _dup2(errwrite, 2)
+
+        # Close pipe fds.  Make sure we don't close the
+        # same fd more than once, or standard fds.
+        closed = set([None])
+        for fd in [p2cread, c2pwrite, errwrite]:
+            if fd not in closed and fd > 2:
+                file_actions.add_close(fd)
+                closed.add(fd)
+
+        gc_was_enabled = gc.isenabled()
+        # FIXME Does this bug apply to posix_spawn version?
+        try:
+            # Disable gc to avoid bug where gc -> file_dealloc ->
+            # write to stderr -> hang.  http://bugs.python.org/issue1336
+            gc.disable()
+            self.pid = posix_spawnp(
+                executable,
+                args,
+                file_actions=file_actions,
+                env=env,
+            )
+        except:
+            if gc_was_enabled:
+                gc.enable()
+            raise
+        finally:
+            for fd in fds_to_close_in_parent:
+                os.close(fd)
+
+        # Parent
+        self._watcher = self._loop.child(self.pid)
+        self._watcher.start(self._on_child, self._watcher)
```

Comment:
I assume you copied this implementation from gevent, but I'd like to point out that it appears to be possible for the code to explode here and never re-enable the garbage collector. Which is sad.